lib/tbot/service_ssh_multiplexer.go

```diff
@@ -0,0 +1,709 @@
+			continue
+		}
+	}
+}
+
+func (s *SSHMultiplexerService) Run(ctx context.Context) (err error) {
+	ctx, span := tracer.Start(
+		ctx,
+		"SSHMultiplexerService/Run",
+	)
+	defer func() { tracing.EndSpan(span, err) }()
+
+	authClient, hostDialer, proxyHost, tshConfig, err := s.setup(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer authClient.Close()
+
+	dest := s.cfg.Destination.(*config.DestinationDirectory)
+	l, err := createListener(
+		ctx,
+		s.log,
+		fmt.Sprintf("unix://%s", filepath.Join(dest.Path, sshMuxSocketName)))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer l.Close()
+
+	eg, egCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		defer context.AfterFunc(egCtx, func() { _ = l.Close() })()
+		for {
+			downstream, err := l.Accept()
+			if err != nil {
+				if utils.IsUseOfClosedNetworkError(err) {
+					return nil
+				}
+
+				s.log.WarnContext(
+					egCtx,
+					"Error encountered accepting connection, sleeping and continuing",
+					"error",
+					err,
+				)
+				select {
+				case <-ctx.Done():
+					return nil
+				case <-time.After(50 * time.Millisecond):
+				}
+
+				continue
+			}
+
+			go func() {
+				muxReqsStartedCounter.Inc()
+				muxReqsInflightGauge.Inc()
+				defer muxReqsInflightGauge.Dec()
+
+				err := s.handleConn(
+					egCtx, tshConfig, authClient, hostDialer, proxyHost, downstream,
+				)
+
+				var status string
+				switch {
+				case utils.IsOKNetworkError(err):
+					// We reduce the verbosity here since these are to be
+					// expected and are not usually indicative of a problem.
+					status = "OK_NETWORK_ERROR"
+					s.log.DebugContext(egCtx, "Handler exited with a network error", "error", err)
+				case err != nil && !errors.Is(err, context.Canceled):
+					status = "ERROR"
+					s.log.WarnContext(egCtx, "Handler exited with error", "error", err)
+				default:
+					status = "OK"
+				}
+				muxReqsHandledCounter.WithLabelValues(status).Inc()
+			}()
+		}
+	})
+	eg.Go(func() error {
+		return s.identityRenewalLoop(egCtx, proxyHost)
+	})
+
+	return eg.Wait()
+}
+
+func (s *SSHMultiplexerService) handleConn(
+	ctx context.Context,
+	tshConfig *libclient.TSHConfig,
+	authClient *authclient.Client,
+	hostDialer *cyclingHostDialClient,
+	proxyHost string,
+	downstream net.Conn,
+) (err error) {
+	ctx, span := tracer.Start(
+		ctx,
+		"SSHMultiplexerService/handleConn",
+		oteltrace.WithNewRoot(),
+	)
+	defer func() { tracing.EndSpan(span, err) }()
+	defer downstream.Close()
+
+	// The first thing downstream will send is the multiplexing request which is
+	// the "[host]:[port]\x00" format.
+	buf := bufio.NewReader(downstream)
+	reqBytes, err := buf.ReadBytes('\x00')
+	if err != nil {
+		return trace.Wrap(err, "reading request")
+	}
+	reqBytes = reqBytes[:len(reqBytes)-1] // Drop the NULL.
+	splits := bytes.Split(reqBytes, []byte{':'})
+	if len(splits) != 2 {
+		return trace.BadParameter("malformed request")
+	}
+	host := string(splits[0])
+	port := string(splits[1])
```

Comment:
IPv6 addresses have plenty of `:`.

```suggestion
	req, err := buf.ReadString('\x00')
	if err != nil {
		return trace.Wrap(err, "reading request")
	}
	req = req[:len(req)-1] // Drop the NUL.
	host, port, err := utils.SplitHostPort(req)
	if err != nil {
		return trace.Wrap(err, "malformed request")
	}
```